stdlib/json.gr

```diff
@@ -0,0 +1,2103 @@
+ * JSON format along with a human readable text message. This can happen when
+ * it contains number values `NaN`, `Infinity` or `-Infinity`.
+ */
+provide enum JsonToStringError {
+  /**
+   * The JSON object contains a number value of `NaN`, `Infinity` or `-Infinity`.
+   */
+  InvalidNumber(String),
+}
+
+/**
+ * Controls how indentation is performed in custom formatting.
+ *
+ * Following examples have whitespaces and line breaks replaced with visible
+ * characters for illustrative purposes.
+ *
+ * `NoIndentation`
+ * ```
+ * {↵
+ * "currency":·"€",↵
+ * "price":·99.9↵
+ * }
+ * ```
+ *
+ * `IndentWithTab`
+ * ```
+ * {↵
+ * →"currency":·"€",↵
+ * →"price":·99.9↵
+ * }
+ * ```
+ *
+ * `IndentWithSpaces(2)`
+ * ```
+ * {↵
+ * ··"currency":·"€",↵
+ * ··"price":·99.9↵
+ * }
+ * ```
+ *
+ * `IndentWithSpaces(4)`
+ * ```
+ * {↵
+ * ····"currency":·"€",↵
+ * ····"price":·99.9↵
+ * }
+ * ```
+ */
+provide enum IndentationFormat {
+  NoIndentation,
+  IndentWithTab,
+  IndentWithSpaces(Number),
+}
+
+/**
+ * Controls how arrays are printed in custom formatting.
+ *
+ * Following examples have whitespaces and line breaks replaced with visible
+ * characters for illustrative purposes.
+ *
+ * `CompactArrayEntries`
+ * ```
+ * []
+ * ```
+ *
+ * ```
+ * [1]
+ * ```
+ *
+ * ```
+ * [1,2,3]
+ * ```
+ *
+ * `SpacedArrayEntries`
+ * ```
+ * [ ]
+ * ```
+ *
+ * ```
+ * [1]
+ * ```
+ *
+ * ```
+ * [1,2,3]
+ * ```
+ *
+ * `OneArrayEntryPerLine`
+ * ```
+ * []
+ * ```
+ *
+ * ```
+ * [↵
+ * ··1↵
+ * ]
+ * ```
+ *
+ * ```
+ * [↵
+ * ··1,↵
+ * ··2,↵
+ * ··3↵
+ * ]
+ * ```
+ */
+provide enum ArrayFormat {
+  CompactArrayEntries,
+  SpacedArrayEntries,
+  OneArrayEntryPerLine,
+}
+
+/**
+ * Controls how objects are printed in custom formatting.
+ *
+ * Following examples have whitespaces and line breaks replaced with visible
+ * characters for illustrative purposes.
+ *
+ * `CompactObjectEntries`
+ * ```
+ * {}
+ * ```
+ *
+ * ```
+ * {"a":1}
+ * ```
+ *
+ * ```
+ * {"a":1,"b":2,"c":3}
+ * ```
+ *
+ * `SpacedObjectEntries`
+ * ```
+ * { }
+ * ```
+ *
+ * ```
+ * {"a": 1}
+ * ```
+ *
+ * ```
+ * {"a": 1, "b": 2, "c": 3}
+ * ```
+ *
+ * `OneObjectEntryPerLine`
+ * ```
+ * {}
+ * ```
+ *
+ * ```
+ * {↵
+ * ··"a":·1↵
+ * }
+ * ```
+ *
+ * ```
+ * {↵
+ * ··"a":·1,↵
+ * ··"b":·2,↵
+ * ··"c":·3↵
+ * }
+ * ```
+ */
+provide enum ObjectFormat {
+  CompactObjectEntries,
+  SpacedObjectEntries,
+  OneObjectEntryPerLine,
+}
+
+/**
+ * Controls line ending type in custom formatting.
+ */
+provide enum LineEnding {
+  NoLineEnding,
+  LineFeed,
+  CarriageReturnLineFeed,
+  CarriageReturn,
+}
+
+/*
+ * Allows fine-grained control of formatting in JSON printing.
+ */
+record FormattingSettings {
+  indentation: IndentationFormat,
+  arrayFormat: ArrayFormat,
+  objectFormat: ObjectFormat,
+  lineEnding: LineEnding,
+  finishWithNewLine: Bool,
+  escapeAllControlPoints: Bool,
+  escapeHTMLUnsafeSequences: Bool,
+  escapeNonASCII: Bool,
+}
```

Comment:
Wasn't this moved to the `Custom` variant record? If so, this should be removed

Reply:
Not quite, we use this internally still so we can standardize the format of settings internally which reduces the complexity.